nengo_ocl/clra_gemv.py

```diff
@@ -1238,43 +1240,376 @@ def choose_plans(self):
 
     Parameters
     ----------
-    A_indices, A_indptr : PyOpenCL array
-        Column sparse row index specifications
-    A_data : PyOpenCL array
-        Matrix values at those indices
+    queue : cl.CommandQueue
+        The queue for this plan.
+    hA : np.array or scipy.sparse.spmat
+        Matrix to multiply in any Numpy or Scipy format. ``Y = hA.dot(X)``
     X, Y : CLRaggedArrays of length 1
         Input/output data.
     inc : bool
-        Whether to increment ``Y`` (True), or set it (False).
+        Whether the plan will increment or set ``Y``.
+    tag : str
+        Tag for the plan.
+    """
+
+    def __init__(
+        self,
+        queue,
+        hA,
+        X,
+        Y,
+        inc=False,
+        tag=None,
+    ):
+        """Initializing will
+            - format the sparse matrix
+            - transfer matrix to device
+            - check the validity of device data
+            - create the plans
+        Most often, the caller will use prog_obj.plans after initializing
+        """
+        self.queue = queue
+        self.hA = hA  # Host matrix in numpy or scipy sparse with methods such as "dot"
+        self.X = X
+        self.Y = Y
+        self.inc = inc
+        self.tag = tag
+
+        self.hAstruct = None  # Host array data. Depends on storage format.
+        self.dAstruct = None  # Device array data. Mirror of hAstruct.
 
-    Notes
-    -----
-    This function crashes when there are >10M nonzero weights. A potential solution
-    would be some way to tell each work item to do multiple rows.
+        self.to_hostdata()
+        self.to_device()
+        self.validate_data()
+        self.plans = self.choose_plans()
+        self.assign_plan_characteristics(self.plans)
+
+    def assign_plan_characteristics(self, plans):
+        pass
+
+    def choose_plans(self):
+        raise NotImplementedError("Subclass must implement")
+
+    def to_device(self):
+        raise NotImplementedError("Subclass must implement")
+
+    def to_hostdata(self):
+        raise NotImplementedError("Subclass must implement")
+
+    def validate_data(self):
+        raise NotImplementedError("Subclass must implement")
+
+
+class plan_ellpack(spmv_prog):
+    """SPMV program with the ELLPACK format.
+
+    This class chooses the implementation based on the size of the problem.
     """
-    assert len(X) == len(Y) == 1
 
-    for arr in [X, Y]:
-        assert (arr.stride1s == 1).all()
-        if not ((arr.shape1s == 1).all() and (arr.stride0s == 1).all()):
-            raise NotImplementedError(
-                "OCL SparseDot only supports matrix-vector currently, not matrix-matrix"
+    @staticmethod
+    def scipy2elldata(scipy_mat, force_nonempty=True):
+        """ Works with scipy sparse or numpy dense argument """
+        scipy_mat = scipy_mat.copy()
+        try:
+            lilmat = scipy_mat.tolil()
+        except AttributeError:
+            lilmat = scipy_sparse.lil_matrix(scipy_mat)
```

Comment:
Is there a reason to not just use this line for all cases? What's the reasoning behind trying `scipy_mat.tolil()` first?

Reply:
Originally this was an attempt to avoid a copy and keep the underlying data. The first one returns the same object. I just looked at scipy, and the second one would keep the same data even though it is a different object. Anyways, it's moot because you don't want to maintain the link sometimes and not other times, depending on input.